homeassistant/helpers/event.py

```diff
@@ -603,29 +607,40 @@ def calculate_next(now: datetime) -> None:
 
     # Make sure rolling back the clock doesn't prevent the timer from
     # triggering.
-    last_now: Optional[datetime] = None
+    cancel_callback: Optional[asyncio.TimerHandle] = None
+    calculate_next(next_time)
 
     @callback
-    def pattern_time_change_listener(event: Event) -> None:
+    def pattern_time_change_listener() -> None:
         """Listen for matching time_changed events."""
-        nonlocal next_time, last_now
+        nonlocal next_time, cancel_callback
 
-        now = event.data[ATTR_NOW]
+        now = pattern_utc_now()
+        hass.async_run_job(action, dt_util.as_local(now) if local else now)
 
-        if last_now is None or now < last_now:
-            # Time rolled back or next time not yet calculated
+        if next_time <= now:
+            calculate_next(now + timedelta(seconds=1))
+        else:
```

Comment:
Why do we still calculate `next` if it is already in the future ?

Reply:
This made sense when the event time was being passed in but I don't think it does anymore

I think we can make this simpler and always set this to `calculate_next(now + timedelta(seconds=1))`

side note: we have to add the second to avoid it firing again right away.